core/src/main/scala/kafka/log/Log.scala

```diff
@@ -507,6 +507,10 @@ class Log(val dir: File,
     // attempt to read beyond the log end offset is an error
     if(startOffset > next || entry == null)
       throw new OffsetOutOfRangeException("Request for offset %d but we only have log segments in the range %d to %d.".format(startOffset, segments.firstKey, next))
+    
+    maxOffset.foreach((offset) => 
```

Comment:
We probably should do this check in ```ReplicaManager.readFromLocalLog()```.

From the original ticket, it is not clear if there is unclean leader election. It seems that the consumer will not fetch beyond high watermark unless the HW rewinds which could happen during unclean leader election.